python/lsst/ctrl/mpexec/execFixupDataId.py

```diff
@@ -0,0 +1,103 @@
+from typing import Any, Iterable, Tuple, Union
+
+from lsst.pipe.base import QuantumIterData
+from .executionGraphFixup import ExecutionGraphFixup
+
+
+class ExecFixupDataId(ExecutionGraphFixup):
+    """Implementation of ExecutionGraphFixup for ordering of tasks based
+    on DataId values.
+
+    This class is a trivial implementation mostly useful as an example,
+    though it can be used to make actual fixup instances by defining
+    a method that instantiates it, e.g.::
+
+        # lsst/ap/verify/ci_fixup.py
+
+        from lsst.ctrl.mpexec.execFixupDataId import ExecFixupDataId
+
+        def assoc_fixup():
+            return ExecFixupDataId(taskLabel="ap_assoc",
+                                   dimensions=("visit", "detector"))
+
+
+    and then executing pipetask::
+
+        pipetask run --graph-fixup=lsst.ap.verify.ci_fixup.assoc_fixup ...
+
+    This will add new dependencies between quanta executed by the task with
+    label "ap_assoc". Quanta with higher visit number will depend on quanta
+    with lower visit number and their execution will wait until lower visit
+    number finishes.
+
+    Parameters
+    ----------
+    taskLabel : `str`
+        The label of the task for which to add dependencies.
+    dimensions : `str` or iterable of `str`
+        One or more dimension names, quanta execution will be ordered
+        according to values of these dimensions.
+    reverse : `bool`, optional
+        If `False` (default) then quanta with higher values of dimensions
+        will be executed after quanta with lower values, otherwise the order
+        is reversed.
+    """
+
+    def __init__(self, taskLabel: str, dimensions: Union[str, Iterable[str]], reverse: bool = False):
+        self.taskLabel = taskLabel
+        self.dimensions = dimensions
+        self.reverse = reverse
+        if isinstance(self.dimensions, str):
+            self.dimensions = [self.dimensions]
+
+    def _key(self, qdata: QuantumIterData) -> Tuple[Any, ...]:
+        """Produce comparison key for quantum data.
+
+        Parameters
+        ----------
+        qdata : `QuantumIterData`
+
+        Returns
+        -------
+        key : `tuple`
+        """
+        dataId = qdata.quantum.dataId
+        key = tuple(dataId[dim] for dim in self.dimensions)
+        return key
+
+    def fixupQuanta(self, quanta: Iterable[QuantumIterData]) -> Iterable[QuantumIterData]:
+        # Docstring inherited from ExecutionGraphFixup.fixupQuanta
+        quanta = list(quanta)
+        taskQuanta = [q for q in quanta if q.taskDef.label == self.taskLabel]
+        if not taskQuanta:
+            raise ValueError(f"Cannot find task with label {self.taskLabel}")
+        # order quanta by dataId dimensions
+        taskQuanta.sort(key=self._key, reverse=self.reverse)
+        for i, qdata in enumerate(taskQuanta):
+            # add previous quanta as dependcency
+            if i > 0:
+                qdata.dependencies |= frozenset([taskQuanta[i-1].index])
```

Comment:
Hmm... won't this mean that even if I sort by `visit` only, all quanta for different detectors but the same visit will still be executed sequentially? That should be fine for current testing (association is a negligible part of our CI-scale running time), but may slow things down as runs get bigger.

Reply:
This is true, I think I can update this implementation to work with multiple quanta per key, hang on.

Reply:
@kfindeisen, I have updated the algorithm, I checked that it works but please review it to see that it also works for you.

Reply:
No objections 👍